core/src/test/java/com/linecorp/armeria/client/HttpClientMaxConcurrentStreamTest.java

```diff
@@ -168,4 +179,230 @@ public void connectionClosed(SessionProtocol protocol, InetSocketAddress remoteA
             assertThat(closes).hasValue(0);
         }
     }
+
+    @Test
+    void handleExceedsMaxStreamsBasicCase() throws Exception {
+        final List<ClientConnectionTimings> connectionTimings = new ArrayList<>();
+        final WebClient client = WebClient.builder(server.uri(SessionProtocol.H2C, "/"))
+                                          .factory(clientFactory)
+                                          .decorator((delegate, ctx, req) -> {
+                                              ctx.logBuilder().whenAvailable(RequestLogProperty.SESSION)
+                                                 .thenAccept(requestLog -> {
+                                                     connectionTimings.add(requestLog.connectionTimings());
+                                                 });
+                                              return delegate.execute(ctx, req);
+                                          })
+                                          .build();
+        final List<CompletableFuture<AggregatedHttpResponse>> receivedResponses = new ArrayList<>();
+        final int numRequests = MAX_CONCURRENT_STREAMS + 1;
+
+        runInsideEventLoop(clientFactory.eventLoopGroup(), () -> {
+            for (int i = 0; i < numRequests; i++) {
+                receivedResponses.add(client.get(PATH).aggregate());
+            }
+        });
+
+        await().untilAsserted(() -> assertThat(responses).hasSize(numRequests));
+        assertThat(connectionTimings.stream().filter(
+                timings -> timings.pendingAcquisitionDurationNanos() > 0))
+                .hasSize(numRequests - 1);
+    }
+
+    @Test
+    void openMinimalConnectionsWhenExceededMaxStreams() throws Exception {
+        final List<ClientConnectionTimings> connectionTimings = new ArrayList<>();
```

Comment:
Could use thread-safe data structure like `ConcurrentLinkedQueue`, because it will be updated from other threads.